### PR TITLE
Colors!

### LINKED ISSUE #2845: Colors!
This is a feature request that I'd *really* like to see before 1.0 hits. Basically: colored output from the compiler. When an error message appears, the word `error` would be red, the word `hint` could be green or purple, and so forth. I can make a PR, though someone would need to try it under Windows (I'm using Ubuntu).

### PR DESCRIPTION
**Pull request by kirbyfan64**
Closes #2845. I seriously doubt this is going to be an instant merge, since this is my first take at tweaking the Nim compiler...

### CODE REVIEW

File: compiler/msgs.nim [hunk 1]
```diff
@@ -683,16 +683,43 @@ proc outWriteln*(s: string) =
   if eStdOut in errorOutputs: writeln(stdout, s)
 
 proc msgWriteln*(s: string) =
-  ## Writes to stdout. If --stdout option is given, writes to stderr instead.
+  ## Writes to stdout. If --stderr option is given, writes to stderr instead.
 
   #if gCmd == cmdIdeTools and optCDebug notin gGlobalOptions: return
 
+  var
+    colorStr = s.split(':')[0]
```
Comment (Araq) [inline compiler/msgs.nim]: Meh I dislike parsing the string which can create all sorts of problems. Instead you need to keep the information somehow.

**Comment by Araq** [comment]
Do we really need this? Can't people use an editor instead that captures the compiler's output, can jump to the offending error line, highlights this stuff on its own? Do we really need to care forever about how software development was done the 70ies?

**Comment by reactormonk** [comment]
@Araq This is a O(1) solution. Making editors parse output is O(n).

**Comment by kirbyfan64** [comment]
@Araq Because not all of us use fancy IDEs.

I can store the info somehow, but I was just trying not to end up refactoring too much stuff in crazy ways.

**Comment by Araq** [comment]
@kirbyfan64 This has nothing to do with "fancy IDEs". It's basic stuff.

**Comment by Araq** [comment]
@reactormonk that's not an argument since we want this editor support everywhere anyway.

**Comment by kirbyfan64** [comment]
@Araq Not necessarily. Out of the "normal" programming text editors I've used (Howl, Acme, and Vim to an extent - Sublime and Atom are big and weird), none even support the line number format that Nim uses. For those of us who do development from the terminal, this is a very nice feature.

**Comment by kirbyfan64** [comment]
Oh, and it's not the 70s software development style. ANSI colors weren't used in compilers again. :)

**Comment by reactormonk** [comment]
Which format is standard for error lines? Maybe switch to that one instead?

**Comment by Araq** [comment]
@kirbyfan64 it doesn't really matter. The Unix way of software develpment is timeless. Never worked, never will. ;-)

**Comment by kirbyfan64** [comment]
@Araq And yet we're too stubborn to change it. ;)

**Comment by kirbyfan64** [comment]
@reactormonk `filename:line:column`.

**Comment by reactormonk** [comment]
@kirbyfan64 where does the error message go?

**Comment by kirbyfan64** [comment]
@reactormonk `filename:line:column: error: <msg>`.

**Comment by Varriount** [comment]
I suppose, if @araq is really against this so much, a compiler wrapper could be written to colorize output.

**Comment by Araq** [comment]
@Varriount I'm not against it this much and will accept some variant of the above patch but it makes me sad how much time people waste in their utterly broken programming environments.

**Comment by kirbyfan64** [comment]
@Araq So...the thing I need to do is not try to re-parse the already built string and just tweak the functions?

**Comment by ozra** [comment]
Araq: Some of us don't always have a choice of the programming environment, in certain situations, if there's any intention for Nim to be widely adopted, that must be accounted for. Your specific setup might work for you, and you might never have to adapt to one you didn't setup yourself?

I any event, I'd like this, but it would be nice if there was a flag to use colors output (--use-colors or whatever).

Also, I second the switch from file(y,x) to file:y:x...

**Comment by kirbyfan64** [comment]
I reimplemented it without the string re-parsing. And used `git commit -a --amend; git push fork --force`. How evil. :)

**Comment by kirbyfan64** [comment]
Beep?

File: compiler/extccomp.nim [hunk 2]
```diff
@@ -13,7 +13,7 @@
 # nim files.
 
 import
-  lists, ropes, os, strutils, osproc, platform, condsyms, options, msgs, crc
+  lists, ropes, os, strutils, osproc, platform, condsyms, options, msgs, crc, terminal
```
Comment (Araq) [inline compiler/extccomp.nim]: extccomp still doesn't require terminal.

Comment (kirbyfan64) [inline compiler/extccomp.nim]: @Araq Fixed.

**Comment by kirbyfan64** [comment]
So *now* I think it's ready.

**Comment by ozra** [comment]
The options doesn't seem to be listed with `nim --help` or `nim --advanced` - had to look at the source..

**Comment by kirbyfan64** [comment]
@ozra Oops. :(

**Pull request merged**